src/ConnectedMode/ServerSentEvents/ServerSentEventSessionManager.cs

```diff
@@ -18,29 +18,104 @@
 using SonarLint.VisualStudio.Core.ServerSentEvents.TaintVulnerabilities;
-using SonarLint.VisualStudio.Core.Binding;
 using SonarQube.Client;
 
 namespace SonarLint.VisualStudio.ConnectedMode.ServerSentEvents
 {
-    internal interface IServerSentEventSessionManager
+    internal interface IServerSentEventSessionManager : IDisposable
     {
     }
 
     [Export(typeof(IServerSentEventSessionManager))]
     [PartCreationPolicy(CreationPolicy.Shared)]
-    internal class ServerSentEventSessionManager : IServerSentEventSessionManager
+    internal sealed class ServerSentEventSessionManager : IServerSentEventSessionManager
     {
+        private readonly IActiveSolutionBoundTracker activeSolutionBoundTracker;
+        private readonly ISonarQubeService sonarQubeClient;
+        private readonly IServerSentEventPump eventPump;
+        private readonly IDisposable issueChangedServerEventSourcePublisher;
+        private readonly IDisposable taintServerEventSourcePublisher;
+        private readonly IThreadHandling threadHandling;
+        private bool disposed;
+        private CancellationTokenSource sessionTokenSource;
+
         [ImportingConstructor]
         public ServerSentEventSessionManager(
             IActiveSolutionBoundTracker activeSolutionBoundTracker,
             ISonarQubeService sonarQubeClient,
+            IServerSentEventPump eventPump,
             IIssueChangedServerEventSourcePublisher issueChangedServerEventSourcePublisher,
-            ITaintServerEventSourcePublisher taintServerEventSourcePublisher)
+            ITaintServerEventSourcePublisher taintServerEventSourcePublisher,
+            IThreadHandling threadHandling)
+        {
+            this.activeSolutionBoundTracker = activeSolutionBoundTracker;
+            this.sonarQubeClient = sonarQubeClient;
+            this.eventPump = eventPump;
+            this.issueChangedServerEventSourcePublisher = issueChangedServerEventSourcePublisher;
+            this.taintServerEventSourcePublisher = taintServerEventSourcePublisher;
+            this.threadHandling = threadHandling;
+            activeSolutionBoundTracker.SolutionBindingChanged += OnSolutionChanged;
+        }
+
+        public void Dispose()
+        {
+            if (disposed)
+            {
+                return;
+            }
+
+            activeSolutionBoundTracker.SolutionBindingChanged -= OnSolutionChanged;
+            sessionTokenSource?.Cancel();
+            issueChangedServerEventSourcePublisher?.Dispose();
+            taintServerEventSourcePublisher?.Dispose();
+            disposed = true;
+        }
+
+        private void OnSolutionChanged(object sender, ActiveSolutionBindingEventArgs activeSolutionBindingEventArgs)
+        {
+            var bindingConfiguration = activeSolutionBindingEventArgs.Configuration;
+            var isInConnectedMode = !bindingConfiguration.Equals(BindingConfiguration.Standalone);
+
+            EndSession();
```

Comment:
I think there might be a race condition problem here. 
1) User opens a solution, sessionTokenSource=null, InitializeSessionAsync sets sessionTokenSource=value
2) User closes the solution, quickly opens another
==> 
sessionTokenSource=previous initialize
session is cancelled, new InitializeSessionAsync is scheduled but not executed yet
3) User closes the solution
==> sessionTokenSource= null because InitializeSessionAsync didn't run it, and no session is cancelled. A thread remains.

We need to handle this with closure, storing the session token created by InitializeSessionAsync

Reply:
I think moving the sessionTokenSource initialization to event handler method to run synchronously would fix this, no?